homeassistant/components/nest/__init__.py

```diff
@@ -173,17 +188,16 @@ def handle_event(self, event_message: EventMessage):
         traits = event_message.resource_update_traits
         if traits:
             _LOGGER.debug("Trait update %s", traits.keys())
+            # This event triggered an update to a device that changed some
+            # properties which the DeviceManager should already have received.
+            # Send a signal to refresh state of all listening devices.
+            async_dispatcher_send(self._hass, SIGNAL_NEST_UPDATE)
         events = event_message.resource_update_events
         if events:
             _LOGGER.debug("Event Update %s", events.keys())
-
-        if not event_message.resource_update_traits:
-            # Note: Currently ignoring events like camera motion
-            return
-        # This event triggered an update to a device that changed some
-        # properties which the DeviceManager should already have received.
-        # Send a signal to refresh state of all listening devices.
-        async_dispatcher_send(self._hass, SIGNAL_NEST_UPDATE)
+            for (event, data) in events.items():
+                if event in EVENT_TRAITS:
+                    self._hass.bus.fire(NEST_EVENT, data)
```

Comment:
Is this method being invoked in async or sync context? You are using `async_dispatcher_send`, so I am assuming async, in that case:
```suggestion
                    self._hass.bus.async_fire(NEST_EVENT, data)
```

If it's sync, it should be `dispatcher_send`. 

Can you confirm that `resource_update_events` includes the `device_id` of the device that fired the event? It's something I added to the documentation last week <https://developers.home-assistant.io/docs/integration_events>. It will help with writing automations as we can listen for `nest_event` with `device_id: <thermostat ID>` to get all events of the thermostat.

Reply:
Thanks for the pointer. 

I've updated this to create a separate data message, independent from the SDM library.  I have two questions here:
1) Since this is from a pubsub background thread, i am pretty sure this is not async.  However, it only passes the tests in async, as the callback isnt fired by the time the test has run.  I was unable to find a lot of tests that exercise the blocking dispatcher_send() or bus.fire() that also had tests.  Have a suggestion for how to block until all events have run?
2) Should this move device updates to use the bus instead of the dispatcher? It'd basically just make both ways to listen to events uniform, and it could also only update specific devices that changed only.

Thanks